ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/ForkChoiceUtil.java

```diff
@@ -294,7 +275,57 @@ public static BlockProcessingRecord on_block(
         > 0) {
       store.setFinalizedCheckpoint(state.getFinalized_checkpoint());
     }
-    return new BlockProcessingRecord(preState, block, state);
+
+    final BlockProcessingRecord record = new BlockProcessingRecord(preState, block, state);
+    return BlockImportResult.successful(record);
+  }
+
+  private static Optional<BlockImportResult> checkOnBlockPreconditions(
+      final BeaconBlock block, final BeaconState preState, final ReadOnlyStore store) {
+    if (preState == null) {
+      return Optional.of(BlockImportResult.FAILED_UNKNOWN_PARENT);
+    }
+    if (blockIsFromFuture(block, preState, store)) {
+      return Optional.of(BlockImportResult.FAILED_BLOCK_IS_FROM_FUTURE);
+    }
+    if (!blockDescendsFromLatestFinalizedBlock(block, store)) {
+      return Optional.of(BlockImportResult.FAILED_INVALID_ANCESTRY);
+    }
+    return Optional.empty();
+  }
+
+  private static boolean blockIsFromFuture(
+      BeaconBlock block, final BeaconState preState, ReadOnlyStore store) {
+    final UnsignedLong genesisTime = preState.getGenesis_time();
+    final UnsignedLong now = store.getTime();
+    final UnsignedLong blockTime = getBlockTime(block, genesisTime);
+    return now.compareTo(blockTime) < 0;
+  }
+
+  private static UnsignedLong getBlockTime(BeaconBlock block, UnsignedLong genesisTime) {
+    return genesisTime.plus(block.getSlot().times(UnsignedLong.valueOf(SECONDS_PER_SLOT)));
+  }
+
+  private static boolean blockDescendsFromLatestFinalizedBlock(
+      final BeaconBlock block, final ReadOnlyStore store) {
+    final Checkpoint finalizedCheckpoint = store.getFinalizedCheckpoint();
+
+    // Make sure this block's slot is after the latest finalized slot
+    final UnsignedLong finalizedEpochStartSlot =
+        compute_start_slot_at_epoch(finalizedCheckpoint.getEpoch());
+    if (block.getSlot().compareTo(finalizedEpochStartSlot) <= 0) {
+      return false;
+    }
+
+    // Make sure this block descends from the finalized block
+    final UnsignedLong finalizedSlot = store.getBlock(finalizedCheckpoint.getRoot()).getSlot();
+    final Bytes32 ancestorAtFinalizedSlot =
+        get_ancestor(store, block.getParent_root(), finalizedSlot);
```

Comment:
Instead of passing the block's root into the `get_ancestor` method, we're now passing the block parent's root.  The only case this affects is when we're importing the finalized block itself.  But we'll never hit this case because the check above will return early if the block's slot is less than or equal to the finalized slot.

This change lets us avoid adding the block to the store before running these checks.